examples/amm/src/contract.rs

```diff
@@ -0,0 +1,342 @@
+    ) -> Result<ExecutionResult<Self::Message>, AmmError> {
+        let mut result = ExecutionResult::default();
+        match operation {
+            Operation::ExecuteOperation { operation } => {
+                if context.chain_id == system_api::current_application_id().creation.chain_id {
+                    self.execute_order_local(operation).await?;
+                } else {
+                    self.execute_order_remote(&mut result, operation).await?;
+                }
+            }
+        }
+
+        Ok(result)
+    }
+
+    async fn execute_message(
+        &mut self,
+        context: &MessageContext,
+        message: Self::Message,
+    ) -> Result<ExecutionResult<Self::Message>, AmmError> {
+        ensure!(
+            context.chain_id == system_api::current_application_id().creation.chain_id,
+            AmmError::AmmChainOnly
+        );
+        match message {
+            Message::ExecuteOperation { operation } => {
+                self.execute_order_local(operation).await?;
+            }
+        }
+        Ok(ExecutionResult::default())
+    }
+
+    async fn handle_application_call(
+        &mut self,
+        context: &CalleeContext,
+        argument: Self::ApplicationCall,
+        _forwarded_sessions: Vec<SessionId>,
+    ) -> Result<ApplicationCallResult<Self::Message, Self::Response, Self::SessionState>, AmmError>
+    {
+        let mut result = ApplicationCallResult::default();
+        match argument {
+            ApplicationCall::ExecuteOperation { operation } => {
+                if context.chain_id == system_api::current_application_id().creation.chain_id {
+                    self.execute_order_local(operation).await?;
+                } else {
+                    self.execute_order_remote(&mut result.execution_result, operation)
+                        .await?;
+                }
+            }
+        }
+
+        Ok(result)
+    }
+
+    async fn handle_session_call(
+        &mut self,
+        _context: &CalleeContext,
+        _session: (),
+        _argument: (),
+        _forwarded_sessions: Vec<SessionId>,
+    ) -> Result<SessionCallResult<Self::Message, Self::Response, Self::SessionState>, AmmError>
+    {
+        Err(AmmError::SessionsNotSupported)
+    }
+}
+
+impl Amm {
+    async fn execute_order_local(&mut self, operation: OperationType) -> Result<(), AmmError> {
+        match operation {
+            OperationType::Swap {
+                owner,
+                input_token_idx,
+                input_amount,
+                output_token_idx,
+            } => {
+                if input_amount == Amount::ZERO {
+                    return Err(AmmError::NoZeroAmounts);
+                }
+
+                if input_token_idx > 1 || output_token_idx > 1 {
+                    return Err(AmmError::InvalidTokenIdx);
+                }
+
+                if input_token_idx == output_token_idx {
+                    return Err(AmmError::EqualTokensError);
+                }
```

Comment:
Are you planning to support more than two tokens soon?
If not, I'd remove `output_tokens_idx` and just make it `1 - input_tokens_idx`.

Reply:
True! Thank you